sklearn/metrics/_classification.py

```diff
@@ -337,62 +340,79 @@ def confusion_matrix(
     (np.int64(0), np.int64(2), np.int64(1), np.int64(1))
     """
     y_true, y_pred = attach_unique(y_true, y_pred)
+    xp, _, device_ = get_namespace_and_device(y_true, y_pred, labels, sample_weight)
     y_type, y_true, y_pred = _check_targets(y_true, y_pred)
     if y_type not in ("binary", "multiclass"):
         raise ValueError("%s is not supported" % y_type)
 
     if labels is None:
         labels = unique_labels(y_true, y_pred)
     else:
-        labels = np.asarray(labels)
+        labels = xp.asarray(labels)
         n_labels = labels.size
         if n_labels == 0:
-            raise ValueError("'labels' should contains at least one label.")
+            raise ValueError("'labels' should contain at least one label.")
         elif y_true.size == 0:
-            return np.zeros((n_labels, n_labels), dtype=int)
-        elif len(np.intersect1d(y_true, labels)) == 0:
+            return xp.zeros((n_labels, n_labels), dtype=xp.int64, device=device_)
+        elif not _isin(labels, y_true, xp=xp).any():
             raise ValueError("At least one label specified must be in y_true")
 
     if sample_weight is None:
-        sample_weight = np.ones(y_true.shape[0], dtype=np.int64)
+        sample_weight = xp.ones(y_true.shape[0], dtype=xp.int64, device=device_)
     else:
-        sample_weight = np.asarray(sample_weight)
+        sample_weight = xp.asarray(sample_weight, device=device_)
 
     check_consistent_length(y_true, y_pred, sample_weight)
 
-    n_labels = labels.size
+    n_labels = size(labels)
     # If labels are not consecutive integers starting from zero, then
     # y_true and y_pred must be converted into index form
     need_index_conversion = not (
-        labels.dtype.kind in {"i", "u", "b"}
-        and np.all(labels == np.arange(n_labels))
-        and y_true.min() >= 0
-        and y_pred.min() >= 0
+        xp.isdtype(labels.dtype, ("signed integer", "unsigned integer", "bool"))
+        and xp.all(labels == xp.arange(n_labels, device=device_))
+        and xp.min(y_true) >= 0
+        and xp.min(y_pred) >= 0
     )
     if need_index_conversion:
-        label_to_ind = {y: x for x, y in enumerate(labels)}
-        y_pred = np.array([label_to_ind.get(x, n_labels + 1) for x in y_pred])
-        y_true = np.array([label_to_ind.get(x, n_labels + 1) for x in y_true])
+        # convert 0D array into scalar type, see https://github.com/data-apis/array-api-strict/issues/109:
+        if xp.isdtype(labels.dtype, ("real floating")):
+            scalar_dtype = float
+        else:
+            scalar_dtype = str
+        label_to_ind = {scalar_dtype(entry): idx for idx, entry in enumerate(labels)}
```

Comment:
```suggestion
        label_to_ind = {entry.__dict__["_array"].item(): idx for idx, entry in enumerate(labels)}
```
I wonder if we should do it like this instead? This might spare us the `if xp.isdtype` check above, if it works.